### ISSUE TITLE
Default language under its own baseurl/{{code}}/

### ISSUE THREAD

**Issue by liuteriadautore** [getzola/zola#1660]
I'm testing Zola for a multi language content website, and I want my default language to lives inside its own subfolder - as the translations - as its the preferred SEO layout (https://www.seopress.org/newsroom/featured-stories/top-10-seo-tips-for-managing-a-multilingual-site/). How can I achieve that? I already tried with the following with no success:

No default language set, than,

```
[languages.en]
generate_feed = true
build_search_index = true

[languages.it]
generate_feed = true
build_search_index = true

[languages.pt]
generate_feed = true
build_search_index = true

```

It still generate mydomain/it and mydomain/pt, yet not mydomain/en

**Comment by Keats** [comment]
It's currently not possible. I need to investigate that claim that it is the preferred layout. If it's true, it should be added as an option to Zola.

**Comment by sirthias** [comment]
I'd like to second this request.
In a true multi-lingual site there is no such thing as a "default language"!
The content simply exists in several languages and the first one the user is presented with when visiting the site is determined by content negotiation via the browsers "Accept-Language" header.
Typically the HTTP server responds to a request to the `base_url` with a 302 response pointing to the "right" language, as it determined it from the "Accept-Language" header.

**Comment by Keats** [comment]
> The content simply exists in several languages and the first one the user is presented with when visiting the site is determined by content negotiation via the browsers "Accept-Language" header.
Typically the HTTP server responds to a request to the base_url with a 302 response pointing to the "right" language, as it determined it from the "Accept-Language" header.

This assumes you have a server you can configure. With static sites it might just be in a s3 bucket so you can't assume that and do need some kind of default language. You can try doing the detection in JS but it still means your site needs to be loaded.
For a static site this means you do mean a homepage and then a homepage per language on top.

**Comment by liuteriadautore** [comment]
It's pretty easy to solve both situations - the need for a default language, and each language in its own suburl - which is to have a redirecting index.html in the root of the site which point to the default language suburl. It is how Hugo solve this, and it's embarassingly easy to setup in a server, s3 bucket, github repo, etc...

Enviado do meu Samsung Mobile da Claro
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by kuon** [comment]
I think both behaviour should be supported. Default at `/` and default at `/<code>`.

My current workaround is to set the default language to something my website does not support.

**Comment by JuliDi** [comment]
Any progress on this?
It is very cumbersome to mess around with one default language and multiple translations.
Would be great to just have each language get its own thing and let the webserver handle the redirect based on headers.

**Comment by Keats** [comment]
> Would be great to just have each language get its own thing and let the webserver handle the redirect based on headers.

Which webserver? Most people deploy their SSG without controlling any webserver.

**Comment by sirthias** [comment]
There are many contexts in which people use SSGs.
One of them is deployment w/o control of the webserver, yes.
But I, for example, have full control over my webserver but still use zola as part of the build process of my site.
It's important to keep an open mind with regard to usage scenarios that are different from one's own.

**Comment by Keats** [comment]
Sure but a new feature needs to work for everyone. Which means here, the home page would be a HTML redirect to base_url/{default_lang}

**Comment by sirthias** [comment]
Yes, I agree that there should be a facility to define a default language which the homepage redirects to.

**Comment by JuliDi** [comment]
> Which webserver? Most people deploy their SSG without controlling any webserver.

I guess there are many scenarios where users can at least control some parts of the webserver. Especially redirects from subdomains etc.
And even though I have full control of my webserver, I still want to use an SSG for ease of use.

> Sure but a new feature needs to work for everyone. Which means here, the home page would be a HTML redirect to base_url/{default_lang}

Maybe the default_language could become optional? So if I do not define a default language, my index.en.md will end up in /en/ and I have to take care of redirecting myself.

For now, I am using @kuon's workaround by setting the default_language to "zz" and have the _index.md redirect to one language (/de/ or /en/). This, of course, will later be done by my webserver that reads the language header and performs the appropriate redirect.

Having to set a default makes it more complicated for me, and presumably also more error-prone: If I mess something up and do not redirect to the appropriate language, I might not even realize it, because the default page somewhat exists (even though I do not want visitors to ever see it).

**Comment by kuon** [comment]
I would like to propose the following solution:

- default language is optional, if present, it would be at `/`, like now
- a new option, perhaps `language.XX.force_output` if set to `true` would emit the site under `/XX` regardless of the default language

If `language.en.force_output = true` and default language is `en`, the english version of the site would be duplicated in the output directory as if they were two different languages. This can be desirable in some situations.

This solution would also not break backward compatibility as `force_output` would default to `false`.

**Comment by Keats** [comment]
> default language is optional, if present

Nope, feeds require a language so it is going to stay as mandatory

> If language.en.force_output = true and default language is en, the english version of the site would be duplicated in the output directory as if they were two different languages. This can be desirable in some situations.

Such as?
From my point of view, there should be an option that just generates HTML redirects to the /xx/ equivalent instead of generating the site twice.

**Comment by kuon** [comment]
There is also the situation where / is a welcome screen with a language picker. So it would be a "language neutral" page

For twice the output, there are some cases where a page can be reached at /en/foo and /foo but it is usually implemented with a dynamic solution in a CRM. It is not my case, but I saw it a few times.


Why default language is required for feed? Feed could be namespaced into /XX too, no?

-- 

Nicolas Goy
https://www.kuon.ch

**Comment by Keats** [comment]
> Why default language is required for feed? Feed could be namespaced into /XX too, no?

Imagine a single language site with no default language.

> For twice the output, there are some cases where a page can be reached at /en/foo and /foo but it is usually implemented with a dynamic solution in a CRM.

Yeah that's what I meant with HTML redirect, the /foo page will automatically redirect to /en/foo

**Comment by JuliDi** [comment]
> Nope, feeds require a language so it is going to stay as mandatory

Well, in the same manner as your arguments: who uses feeds? If I don't use it, it must be a useless feature and does not deserve to be present in zola. Hence, feeds should be deleted to make allow for making the default language optional.

Honestly: if you have good reasons against implementing the feature in some way, I highly appreciate the feedback. But just saying "no" without any alternatives for users facing trouble is not helpful in any way.

I do like how Zola works and I do understand some limitations it has. But I do not understand why you keep blocking every attempt to fix this issue instead of helping to find a solution.

**Comment by Keats** [comment]
>  But I do not understand why you keep blocking every attempt to fix this issue instead of helping to find a solution.

Where did I say I would block it? I just described how it would look 2 messages up...?

**Comment by kuon** [comment]
The option to force the namespacing of the default language and have HTML redirects is fine to me, I made other proposition because I thought it was not liked.

Maybe we could handle both HTTP and HTML redirect with option that can have three values like:

- default -> like now, rendered at /
- HTTP redirect -> rendered at /XX, nothing rendered at /
- HTML redirect -> rendered at  /XX, and /XX/Y is also rendered at /Y as a minimal HTML redirect file (meta header) to /XX/Y if the server cannot be configured for HTTP redirect

Having only the HTML redirect option would also work and the user would be free to configure the HTTP server, this would ensure a proper fallback and make the site more portable.

For default language, it can stay mandatory if the above is implemented. The main point is to be able to have /XX where XX is default language.

**Comment by LitoMore** [comment]
We have `page.lang` for pages. But we can get the `lang` for `base_url/<code>`. It would be good if we can access the `lang` at the root level as well.

Here is my current workaround:

```
{%- if paginator %}
{%- set site_lang = paginator.pages | first() | get(key="lang") %}
{%- set is_zh = site_lang == 'zh' %}
{%- else %}
{%- set is_zh = false %}
{%- endif %}
```

**Comment by Keats** [comment]
All templates should set a `lang` variable in the context (except 404)

**Comment by axelkar** [comment]
Sorry, but is there any progress on this?

**Comment by zlfn** [comment]
Hello, I need this feature as well. I've currently added JavaScript on the default domain to automatically redirect users to the language matching their browser settings (https://zlfn.space).

While non-default English and Japanese versions can be linked directly via `/en/` and `/ja/` prefixes, there's no equivalent `/ko/` path for the default language. I've worked around it with a hacky solution (special-casing the default language in JS before redirecting), but it would be great to have this feature built in.

**Comment by Keats** [comment]
I think it would be ok to add a config file option to enable that?
